File: java/L200/LeetCode0228.java
package L200;

import java.util.ArrayList;
import java.util.List;

public class LeetCode0228 {
    public List<String> summaryRanges(int[] nums) {
        List<String> res = new ArrayList<>();
        if (nums == null || nums.length == 0) {
            return res;
        }
        int left = nums[0];
        int right = nums[0];
        for (int i = 1;i < nums.length;++i) {
            if (nums[i] == right) {
                continue;
            } else if (nums[i] == right + 1) {
                right = nums[i];
            } else {
                add(res, left, right);
                left = nums[i];
                right = nums[i];
            }
        }
        add(res, left, right);
        return res;
    }

    private void add(List<String> res, int left, int right) {
        if (left == right) {
            res.add(String.valueOf(right));
        } else {
            StringBuilder range = new StringBuilder();
            range.append(left);
            range.append("->");
            range.append(right);
            res.add(range.toString());
        }
    }
}

File: cpp/L200/LeetCode0228.cc
#include <algorithm>
#include <vector>
using std::vector;

#include <string>
using std::string;

class Solution {
public:
    vector<string> summaryRanges(vector<int>& nums) {
        vector<string> r;
        int nz = nums.size();
        for (int i = 0;i < nz;) {
            int j{i};
            while (j + 1 < nz && nums[j + 1] == nums[j] + 1) {
                ++j;
            }
            if (j == i) {
                r.push_back(std::to_string(nums[i]));
            } else {
                r.push_back(std::to_string(nums[i]) + "->" + std::to_string(nums[j]));
            }
            i = j + 1;
        }
        return r;
    }
};